// @chakra-ui/hooks/dist/cjs/use-outside-click.js
"use strict";

exports.__esModule = true;
exports.useOutsideClick = useOutsideClick;

var _react = require("react");

var _useLatestRef = require("./use-latest-ref");

/**
 * Example, used in components like Dialogs and Popovers so they can close
 * when a user clicks outside them.
 */
function useOutsideClick(props) {
  var ref = props.ref,
      handler = props.handler;
  var savedHandler = (0, _useLatestRef.useLatestRef)(handler);
  var stateRef = (0, _react.useRef)({
    isPointerDown: false,
    ignoreEmulatedMouseEvents: false
  });
  var state = stateRef.current;
  (0, _react.useEffect)(function () {
    var onPointerDown = function onPointerDown(e) {
      if (isValidEvent(e, ref)) {
        state.isPointerDown = true;
      }
    };

    var onMouseUp = function onMouseUp(event) {
      if (state.ignoreEmulatedMouseEvents) {
        state.ignoreEmulatedMouseEvents = false;
        return;
      }

      if (state.isPointerDown && handler && isValidEvent(event, ref)) {
        state.isPointerDown = false;
        savedHandler.current == null ? void 0 : savedHandler.current(event);
      }
    };

    var onTouchEnd = function onTouchEnd(event) {
      state.ignoreEmulatedMouseEvents = true;

      if (handler && state.isPointerDown && isValidEvent(event, ref)) {
        state.isPointerDown = false;
        savedHandler.current == null ? void 0 : savedHandler.current(event);
      }
    };

    document.addEventListener("mousedown", onPointerDown, true);
    document.addEventListener("mouseup", onMouseUp, true);
    document.addEventListener("touchstart", onPointerDown, true);
    document.addEventListener("touchend", onTouchEnd, true);
    return function () {
      document.removeEventListener("mousedown", onPointerDown, true);
      document.removeEventListener("mouseup", onMouseUp, true);
      document.removeEventListener("touchstart", onPointerDown, true);
      document.removeEventListener("touchend", onTouchEnd, true);
    };
  }, [handler, ref, state.ignoreEmulatedMouseEvents, state.isPointerDown]);
}

function isValidEvent(event, ref) {
  var target = event.target;
  if (event.button > 0) return false; // if the event target is no longer in the document

  if (target) {
    var ownerDocument = target.ownerDocument;

    if (!ownerDocument || !ownerDocument.body.contains(target)) {
      return false;
    }
  }

  return ref.current && !ref.current.contains(target);
}
