// @chakra-ui/portal/dist/cjs/portal.js
"use strict";

exports.__esModule = true;
exports.Portal = void 0;

var _hooks = require("@chakra-ui/hooks");

var _utils = require("@chakra-ui/utils");

var React = _interopRequireWildcard(require("react"));

var ReactDOM = _interopRequireWildcard(require("react-dom"));

var _portalManager = require("./portal-manager");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var _createContext = (0, _utils.createContext)({
  strict: false,
  name: "PortalContext"
}),
    PortalContextProvider = _createContext[0],
    usePortalContext = _createContext[1];

/**
 * Portal
 *
 * Declarative component used to render children into a DOM node
 * that exists outside the DOM hierarchy of the parent component.
 *
 * @see Docs https://chakra-ui.com/components/portal
 */
var Portal = function Portal(props) {
  var onMount = props.onMount,
      onUnmount = props.onUnmount,
      children = props.children,
      getContainer = props.getContainer;
  /**
   * Generate the portal's dom node. We'll wrap the children
   * in this dom node before mounting it.
   */

  var _React$useState = React.useState(function () {
    if (_utils.isBrowser) {
      var div = document.createElement("div");
      div.className = "chakra-portal";
      return div;
    } // for ssr


    return null;
  }),
      portal = _React$useState[0];
  /**
   * This portal might be nested in another portal.
   * Let's read from the portal context to check this.
   */


  var parentPortal = usePortalContext();
  /**
   * If there's a PortalManager rendered, let's read from it.
   * We use the portal manager to manage multiple portals
   */

  var manager = (0, _portalManager.usePortalManager)();
  var append = React.useCallback(function (container) {
    // if user specified a mount node, do nothing.
    if (!portal || !container) return; // else, simply append component to the portal node

    container.appendChild(portal);
  }, [portal]);
  (0, _hooks.useSafeLayoutEffect)(function () {
    var _ref, _ref2;

    // get the custom container from the container prop
    var customContainer = getContainer == null ? void 0 : getContainer();
    /**
     * We need to know where to mount this portal, we have 4 options:
     * - If a mountRef is specified, we'll use that as the container
     * - If portal is nested, use the parent portal node as container.
     * - If it is not nested, use the manager's node as container
     * - else use document.body as containers
     */

    var container = (_ref = (_ref2 = customContainer != null ? customContainer : parentPortal) != null ? _ref2 : manager == null ? void 0 : manager.node) != null ? _ref : document.body;
    /**
     * Append portal node to the computed container
     */

    append(container);
    onMount == null ? void 0 : onMount();
    return function () {
      onUnmount == null ? void 0 : onUnmount();
      if (!portal) return;

      if (container == null ? void 0 : container.contains(portal)) {
        container == null ? void 0 : container.removeChild(portal);
      }
    };
  }, [getContainer, portal, parentPortal, onMount, onUnmount, manager == null ? void 0 : manager.node, append]);
  var portalChildren = (manager == null ? void 0 : manager.zIndex) ? /*#__PURE__*/React.createElement("div", {
    className: "chakra-portal-zIndex",
    style: {
      position: "absolute",
      zIndex: manager.zIndex,
      width: "100%"
    }
  }, children) : children;

  if (!portal) {
    return /*#__PURE__*/React.createElement(React.Fragment, null, portalChildren);
  }

  return /*#__PURE__*/ReactDOM.createPortal( /*#__PURE__*/React.createElement(PortalContextProvider, {
    value: portal
  }, portalChildren), portal);
};

exports.Portal = Portal;

if (_utils.__DEV__) {
  Portal.displayName = "Portal";
}
